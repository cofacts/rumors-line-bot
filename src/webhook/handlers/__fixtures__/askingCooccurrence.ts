import {
  ListArticlesInInitStateQuery,
  ListArticlesInProcessMediaQuery,
} from 'typegen/graphql';

export const textMessage = {
  id: 'text-message',
  text: 'This is a text message',
  articleType: 'TEXT',
  attachmentUrl: null,
  replyCount: 0,
} as const satisfies NonNullable<
  ListArticlesInInitStateQuery['ListArticles']
>['edges'][0]['node'];

export const imageMessage = {
  id: 'img-message',
  articleType: 'TEXT',
  attachmentUrl: 'https://example.com/image.jpg',
  replyCount: 0,
} as const satisfies NonNullable<
  ListArticlesInProcessMediaQuery['ListArticles']
>['edges'][0]['node'];
