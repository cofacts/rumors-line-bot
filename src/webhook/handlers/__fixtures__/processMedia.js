export const oneIdenticalVideoArticle = {
  data: {
    ListArticles: {
      edges: [
        {
          score: 2,
          node: {
            articleType: 'VIDEO',
            attachmentHash: 'ffff8000',
            attachmentUrl: 'http://foo/video.mp4',
            id: 'video-article-1',
          },
        },
      ],
    },
  },
};

export const getVideoArticle = {
  data: {
    GetArticle: {
      text: '',
      replyCount: 2,
      articleReplies: [
        {
          reply: {
            id: 'AVygFA0RyCdS-nWhuaXY',
            type: 'RUMOR',
            text: '查證這則謠言的過程中，約翰走鹿也發現其實網路早就有許多解答，就讓約翰走鹿來借花獻佛一下囉。多家媒體都有針對這則「吃生魚片會有海獸胃腺蟲附著胃壁」的謠言請教醫師，受訪的醫師們皆表示，這則謠言太誇大，海獸胃腺蟲沒那麼神通廣大。\n醫師們表示，台灣的衛生條件很進步，吃生魚片被寄生蟲寄生的案例很少，大部份都是生魚片不乾淨，感染沙門氏菌、大腸桿菌的案例，此外，寄生蟲也不是什麼難治之症，服用一些抗寄生蟲藥就可以殺死胃裡的這些蟲子，除非這些蟲子跑到身體其他部位那才比較難處理，但那也是要大量地吃生魚片（一週兩次以上），中獎機率才會比較高。',
          },
          positiveFeedbackCount: 14,
          negativeFeedbackCount: 2,
        },
        {
          reply: {
            id: 'AVy6LkWIyCdS-nWhuaqu',
            type: 'RUMOR',
            text: '過去台灣只有一個海獸胃腺蟲鑽入胃壁的病例，而且僅有一隻蟲。而經查詢，台北馬偕醫院也沒有名為林清風的醫師。',
          },
          positiveFeedbackCount: 24,
          negativeFeedbackCount: 0,
        },
      ],
    },
  },
};

export const notFound = {
  data: {
    ListArticles: {
      edges: [],
    },
  },
};
