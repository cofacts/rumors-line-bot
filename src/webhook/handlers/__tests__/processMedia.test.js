jest.mock('src/lib/gql');
jest.mock('src/lib/ga');
jest.mock('src/lib/detectDialogflowIntent');

import MockDate from 'mockdate';
import processMedia from '../processMedia';
import * as apiListArticlesResult from '../__fixtures__/processMedia';
import * as apiGetArticleResult from '../__fixtures__/choosingArticle';
import gql from 'src/lib/gql';
import ga from 'src/lib/ga';

beforeEach(() => {
  ga.clearAllMocks();
  gql.__reset();
});

it('one identical article found and choose for user', async () => {
  gql.__push(apiListArticlesResult.oneIdenticalImageArticle);
  gql.__push(apiGetArticleResult.oneImageArticle);

  const data = {
    sessionId: 1497994017447,
  };
  const event = {
    type: 'message',
    timestamp: 1497994016356,
    messageId: '6270464463537',
    message: {
      type: 'image',
      id: '6270464463537',
    },
  };
  const userId = 'Uc76d8ae9ccd1ada4f06c4e1515d46466';
  MockDate.set('2020-01-01');
  expect(await processMedia(data, event, userId)).toMatchSnapshot();
  MockDate.reset();
  expect(gql.__finished()).toBe(true);
  expect(ga.eventMock.mock.calls).toMatchInlineSnapshot(`
    Array [
      Array [
        Object {
          "ea": "MessageType",
          "ec": "UserInput",
          "el": "image",
        },
      ],
      Array [
        Object {
          "ea": "ArticleSearch",
          "ec": "UserInput",
          "el": "ArticleFound",
        },
      ],
      Array [
        Object {
          "ea": "Search",
          "ec": "Article",
          "el": "image-article-1",
          "ni": true,
        },
      ],
      Array [
        Object {
          "ea": "Selected",
          "ec": "Article",
          "el": "image-article-1",
        },
      ],
      Array [
        Object {
          "ea": "Search",
          "ec": "Reply",
          "el": "AVygFA0RyCdS-nWhuaXY",
          "ni": true,
        },
      ],
      Array [
        Object {
          "ea": "Search",
          "ec": "Reply",
          "el": "AVy6LkWIyCdS-nWhuaqu",
          "ni": true,
        },
      ],
    ]
  `);
  expect(ga.sendMock).toHaveBeenCalledTimes(2);
});

it('one article found (not identical)', async () => {
  gql.__push(apiListArticlesResult.oneImageArticle);

  const data = {
    sessionId: 1497994017447,
  };
  const event = {
    type: 'message',
    timestamp: 1497994016356,
    messageId: '6270464463537',
    message: {
      type: 'image',
      id: '6270464463537',
    },
  };
  const userId = 'Uc76d8ae9ccd1ada4f06c4e1515d46466';
  MockDate.set('2020-01-01');
  expect(await processMedia(data, event, userId)).toMatchSnapshot();
  MockDate.reset();
  expect(gql.__finished()).toBe(true);
  expect(ga.eventMock.mock.calls).toMatchInlineSnapshot(`
    Array [
      Array [
        Object {
          "ea": "MessageType",
          "ec": "UserInput",
          "el": "image",
        },
      ],
      Array [
        Object {
          "ea": "ArticleSearch",
          "ec": "UserInput",
          "el": "ArticleFound",
        },
      ],
      Array [
        Object {
          "ea": "Search",
          "ec": "Article",
          "el": "image-article-1",
          "ni": true,
        },
      ],
    ]
  `);
  expect(ga.sendMock).toHaveBeenCalledTimes(1);
});

it('twelve articles found', async () => {
  gql.__push(apiListArticlesResult.twelveImageArticles);

  const data = {
    sessionId: 1497994017447,
  };
  const event = {
    type: 'message',
    timestamp: 1497994016356,
    messageId: '6530038889933',
    message: {
      type: 'image',
      id: '6530038889933',
    },
  };
  const userId = 'Uc76d8ae9ccd1ada4f06c4e1515d46466';

  MockDate.set('2020-01-01');
  const result = await processMedia(data, event, userId);
  MockDate.reset();
  expect(result).toMatchSnapshot();
  expect(gql.__finished()).toBe(true);
  expect(result.replies.length).toBeLessThanOrEqual(5); // Reply message API limit
  const carousel = result.replies.find(({ type }) => type === 'flex').contents;
  expect(carousel.type).toBe('carousel');
  expect(carousel.contents.length).toBeLessThanOrEqual(10); // Flex message carousel 10 bubble limit
  expect(JSON.stringify(carousel).length).toBeLessThan(50 * 1000); // Flex message carousel 50K limit
});

it('should handle image not found', async () => {
  gql.__push(apiListArticlesResult.notFound);
  const data = {
    sessionId: 1497994017447,
  };
  const event = {
    type: 'message',
    timestamp: 1497994016356,
    messageId: '6530038889933',
    message: {
      type: 'image',
      id: '6530038889933',
    },
  };
  const userId = 'Uc76d8ae9ccd1ada4f06c4e1515d46466';
  MockDate.set('2020-01-01');
  expect(await processMedia(data, event, userId)).toMatchSnapshot();
  MockDate.reset();
  expect(gql.__finished()).toBe(true);
  expect(ga.eventMock.mock.calls).toMatchInlineSnapshot(`
    Array [
      Array [
        Object {
          "ea": "MessageType",
          "ec": "UserInput",
          "el": "image",
        },
      ],
      Array [
        Object {
          "ea": "ArticleSearch",
          "ec": "UserInput",
          "el": "ArticleNotFound",
        },
      ],
    ]
  `);
  expect(ga.sendMock).toHaveBeenCalledTimes(1);
});
