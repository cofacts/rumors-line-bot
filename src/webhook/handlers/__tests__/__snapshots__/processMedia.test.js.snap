// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`one article found (not identical) 1`] = `
Object {
  "context": Object {
    "msgs": Array [
      Object {
        "id": "6270464463537",
        "type": "image",
      },
    ],
    "sessionId": 1577836800000,
  },
  "replies": Array [
    Object {
      "altText": "🔍 There are some messages that looks similar to the one you have sent to me.",
      "contents": Object {
        "body": Object {
          "contents": Array [
            Object {
              "text": "🔍 There are some messages that looks similar to the one you have sent to me.",
              "type": "text",
              "wrap": true,
            },
          ],
          "layout": "vertical",
          "type": "box",
        },
        "type": "bubble",
      },
      "type": "flex",
    },
    Object {
      "altText": "Internet rumors are often mutated and shared.
Please choose the version that looks the most similar👇",
      "contents": Object {
        "body": Object {
          "contents": Array [
            Object {
              "text": "Internet rumors are often mutated and shared.
Please choose the version that looks the most similar👇",
              "type": "text",
              "wrap": true,
            },
          ],
          "layout": "vertical",
          "type": "box",
        },
        "type": "bubble",
      },
      "type": "flex",
    },
    Object {
      "altText": "Please choose the most similar message from the list.",
      "contents": Object {
        "contents": Array [
          Object {
            "body": undefined,
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":\\"image-article-1\\",\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_ARTICLE\\"}",
                    "displayText": "I choose ①",
                    "label": "Choose this one",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "① Looks 87% similar",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "md",
              "spacing": "sm",
              "type": "box",
            },
            "hero": Object {
              "size": "full",
              "type": "image",
              "url": "http://foo/image.jpeg",
            },
            "type": "bubble",
          },
          Object {
            "body": Object {
              "contents": Array [
                Object {
                  "flex": 0,
                  "gravity": "top",
                  "maxLines": 5,
                  "text": "None of these messages matches mine :(",
                  "type": "text",
                  "weight": "regular",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "margin": "none",
              "spacing": "none",
              "type": "box",
            },
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":\\"__NO_ARTICLE_FOUND__\\",\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_ARTICLE\\"}",
                    "displayText": "None of these messages matches mine :(",
                    "label": "Tell us more",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "color": "#AAAAAA",
                  "margin": "none",
                  "size": "sm",
                  "text": "😶",
                  "type": "text",
                  "weight": "bold",
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "none",
              "type": "box",
            },
            "type": "bubble",
          },
        ],
        "type": "carousel",
      },
      "type": "flex",
    },
  ],
}
`;

exports[`one identical article found and choose for user 1`] = `
Object {
  "context": Object {
    "msgs": Array [
      Object {
        "id": "6270464463537",
        "type": "image",
      },
    ],
    "sessionId": 1577836800000,
  },
  "replies": Array [
    Object {
      "text": "👨‍👩‍👧‍👦  Volunteer editors have published several replies to this message.

2 of them say it ❌ contains misinformation.",
      "type": "text",
    },
    Object {
      "text": "Let's pick one 👇",
      "type": "text",
    },
    Object {
      "altText": "Please take a look at the following replies.",
      "contents": Object {
        "contents": Array [
          Object {
            "body": Object {
              "contents": Array [
                Object {
                  "align": "start",
                  "margin": "md",
                  "maxLines": 10,
                  "text": "查證這則謠言的過程中，約翰走鹿也發現其實網路早就有許多解答，就讓約翰走鹿來借花獻佛一下囉。多家媒體都有針對這則「吃生魚片會有海獸胃腺蟲附著胃壁」的謠言請教醫師，受訪的醫師們皆表示，這則謠言太誇大，海獸胃腺蟲沒那麼神通廣大。
醫師們表示，台灣的衛生條件很進步，吃生魚片被寄生蟲寄生的案例很少，大部份都是生魚片不乾淨，感染沙門氏菌、大腸桿菌的案例，此外，寄生蟲也不是什麼難治之症，服用一些抗寄生蟲藥就可以殺死胃裡的這些蟲子，除非這些蟲子跑到身體其他部位那才比較難處理，但那也是要大量地吃生魚片（一週兩次以上），中獎機率才會比較高。",
                  "type": "text",
                  "wrap": true,
                },
                Object {
                  "type": "filler",
                },
                Object {
                  "margin": "md",
                  "type": "separator",
                },
                Object {
                  "contents": Array [
                    Object {
                      "size": "xs",
                      "text": "👍 14 users consider this helpful
😕 2 users consider this not useful",
                      "type": "text",
                      "wrap": true,
                    },
                  ],
                  "layout": "horizontal",
                  "margin": "md",
                  "spacing": "none",
                  "type": "box",
                },
              ],
              "layout": "vertical",
              "type": "box",
            },
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":{\\"a\\":\\"image-article-1\\",\\"r\\":\\"AVygFA0RyCdS-nWhuaXY\\"},\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_REPLY\\"}",
                    "displayText": "I choose “查證這則謠言的過程中，約翰走鹿也發現其實網路早⋯⋯”",
                    "label": "👀 Take a look",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "vertical",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "flex": 0,
                  "text": "💬",
                  "type": "text",
                },
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "Someone thinks it contains misinformation",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "none",
              "spacing": "md",
              "type": "box",
            },
            "type": "bubble",
          },
          Object {
            "body": Object {
              "contents": Array [
                Object {
                  "align": "start",
                  "margin": "md",
                  "maxLines": 10,
                  "text": "過去台灣只有一個海獸胃腺蟲鑽入胃壁的病例，而且僅有一隻蟲。而經查詢，台北馬偕醫院也沒有名為林清風的醫師。",
                  "type": "text",
                  "wrap": true,
                },
                Object {
                  "type": "filler",
                },
                Object {
                  "margin": "md",
                  "type": "separator",
                },
                Object {
                  "contents": Array [
                    Object {
                      "size": "xs",
                      "text": "👍 24 users consider this helpful",
                      "type": "text",
                      "wrap": true,
                    },
                  ],
                  "layout": "horizontal",
                  "margin": "md",
                  "spacing": "none",
                  "type": "box",
                },
              ],
              "layout": "vertical",
              "type": "box",
            },
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":{\\"a\\":\\"image-article-1\\",\\"r\\":\\"AVy6LkWIyCdS-nWhuaqu\\"},\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_REPLY\\"}",
                    "displayText": "I choose “過去台灣只有一個海獸胃腺蟲鑽入胃壁的病例，而且⋯⋯”",
                    "label": "👀 Take a look",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "vertical",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "flex": 0,
                  "text": "💬",
                  "type": "text",
                },
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "Someone thinks it contains misinformation",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "none",
              "spacing": "md",
              "type": "box",
            },
            "type": "bubble",
          },
        ],
        "type": "carousel",
      },
      "type": "flex",
    },
  ],
}
`;

exports[`one identical image and similar text found 1`] = `
Object {
  "context": Object {
    "msgs": Array [
      Object {
        "id": "6270464463537",
        "type": "image",
      },
    ],
    "sessionId": 1577836800000,
  },
  "replies": Array [
    Object {
      "altText": "🔍 There are some messages that looks similar to the one you have sent to me.",
      "contents": Object {
        "body": Object {
          "contents": Array [
            Object {
              "text": "🔍 There are some messages that looks similar to the one you have sent to me.",
              "type": "text",
              "wrap": true,
            },
          ],
          "layout": "vertical",
          "type": "box",
        },
        "type": "bubble",
      },
      "type": "flex",
    },
    Object {
      "altText": "Internet rumors are often mutated and shared.
Please choose the version that looks the most similar👇",
      "contents": Object {
        "body": Object {
          "contents": Array [
            Object {
              "text": "Internet rumors are often mutated and shared.
Please choose the version that looks the most similar👇",
              "type": "text",
              "wrap": true,
            },
          ],
          "layout": "vertical",
          "type": "box",
        },
        "type": "bubble",
      },
      "type": "flex",
    },
    Object {
      "altText": "Please choose the most similar message from the list.",
      "contents": Object {
        "contents": Array [
          Object {
            "body": undefined,
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":\\"image-article-1\\",\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_ARTICLE\\"}",
                    "displayText": "I choose ①",
                    "label": "Choose this one",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "① Looks 100% similar",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "md",
              "spacing": "sm",
              "type": "box",
            },
            "hero": Object {
              "size": "full",
              "type": "image",
              "url": "http://foo/image.jpeg",
            },
            "type": "bubble",
          },
          Object {
            "body": Object {
              "contents": Array [
                Object {
                  "contents": Array [
                    Object {
                      "text": "Text matching ",
                      "type": "span",
                    },
                    Object {
                      "color": "#ffb600",
                      "text": "query transcript",
                      "type": "span",
                      "weight": "bold",
                    },
                  ],
                  "flex": 0,
                  "gravity": "top",
                  "maxLines": 12,
                  "type": "text",
                  "weight": "regular",
                  "wrap": true,
                },
              ],
              "layout": "vertical",
              "type": "box",
            },
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":\\"text-article-1\\",\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_ARTICLE\\"}",
                    "displayText": "I choose ②",
                    "label": "Choose this one",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "② Contains relevant text",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "md",
              "spacing": "sm",
              "type": "box",
            },
            "hero": undefined,
            "type": "bubble",
          },
        ],
        "type": "carousel",
      },
      "type": "flex",
    },
  ],
}
`;

exports[`should handle image not found 1`] = `
Object {
  "context": Object {
    "msgs": Array [
      Object {
        "id": "6530038889933",
        "type": "image",
      },
    ],
    "sessionId": 1577836800000,
  },
  "replies": Array [
    Object {
      "altText": "Unfortunately, I currently don’t recognize this message, but I would still like to help.
Do you want someone to fact-check this message?",
      "contents": Object {
        "body": Object {
          "contents": Array [
            Object {
              "text": "Unfortunately, I currently don’t recognize this message, but I would still like to help.
Do you want someone to fact-check this message?",
              "type": "text",
              "wrap": true,
            },
          ],
          "layout": "vertical",
          "type": "box",
        },
        "type": "bubble",
      },
      "type": "flex",
    },
    Object {
      "altText": "Be the first to report the message",
      "contents": Object {
        "body": Object {
          "contents": Array [
            Object {
              "contents": Array [
                Object {
                  "text": "Currently we don’t have this message in our database. If you think it is most likely a rumor, ",
                  "type": "span",
                },
                Object {
                  "color": "#ffb600",
                  "text": "press “🆕 Report to database” to make this message public on Cofacts database ",
                  "type": "span",
                  "weight": "bold",
                },
                Object {
                  "text": "and have volunteers fact-check it. This way you can help the people who receive the same message in the future.",
                  "type": "span",
                },
              ],
              "type": "text",
              "wrap": true,
            },
          ],
          "layout": "vertical",
          "paddingAll": "lg",
          "spacing": "md",
          "type": "box",
        },
        "footer": Object {
          "contents": Array [
            Object {
              "action": Object {
                "data": "{\\"input\\":\\"__POSTBACK_YES__\\",\\"sessionId\\":1577836800000,\\"state\\":\\"ASKING_ARTICLE_SUBMISSION_CONSENT\\"}",
                "displayText": "🆕 Report to database",
                "label": "🆕 Report to database",
                "type": "postback",
              },
              "color": "#ffb600",
              "style": "primary",
              "type": "button",
            },
            Object {
              "action": Object {
                "data": "{\\"input\\":\\"__POSTBACK_NO__\\",\\"sessionId\\":1577836800000,\\"state\\":\\"ASKING_ARTICLE_SUBMISSION_CONSENT\\"}",
                "displayText": "Don’t report",
                "label": "Don’t report",
                "type": "postback",
              },
              "color": "#333333",
              "style": "primary",
              "type": "button",
            },
          ],
          "layout": "vertical",
          "spacing": "sm",
          "type": "box",
        },
        "styles": Object {
          "body": Object {
            "separator": true,
          },
        },
        "type": "bubble",
      },
      "type": "flex",
    },
  ],
}
`;

exports[`twelve articles found 1`] = `
Object {
  "context": Object {
    "msgs": Array [
      Object {
        "id": "6530038889933",
        "type": "image",
      },
    ],
    "sessionId": 1577836800000,
  },
  "replies": Array [
    Object {
      "text": "🔍 There are some messages that looks similar to the one you have sent to me.",
      "type": "text",
    },
    Object {
      "text": "Internet rumors are often mutated and shared.
Please choose the version that looks the most similar👇",
      "type": "text",
    },
    Object {
      "altText": "Please choose the most similar message from the list.",
      "contents": Object {
        "contents": Array [
          Object {
            "body": undefined,
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":\\"image-article-1\\",\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_ARTICLE\\"}",
                    "displayText": "I choose ①",
                    "label": "Choose this one",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "① Looks 100% similar",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "md",
              "spacing": "sm",
              "type": "box",
            },
            "hero": Object {
              "size": "full",
              "type": "image",
              "url": "http://foo/image.jpeg",
            },
            "type": "bubble",
          },
          Object {
            "body": undefined,
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":\\"image-article-2\\",\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_ARTICLE\\"}",
                    "displayText": "I choose ②",
                    "label": "Choose this one",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "② Looks 87% similar",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "md",
              "spacing": "sm",
              "type": "box",
            },
            "hero": Object {
              "size": "full",
              "type": "image",
              "url": "http://foo/image2.jpeg",
            },
            "type": "bubble",
          },
          Object {
            "body": undefined,
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":\\"image-article-1\\",\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_ARTICLE\\"}",
                    "displayText": "I choose ③",
                    "label": "Choose this one",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "③ Looks 100% similar",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "md",
              "spacing": "sm",
              "type": "box",
            },
            "hero": Object {
              "size": "full",
              "type": "image",
              "url": "http://foo/image.jpeg",
            },
            "type": "bubble",
          },
          Object {
            "body": undefined,
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":\\"image-article-2\\",\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_ARTICLE\\"}",
                    "displayText": "I choose ④",
                    "label": "Choose this one",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "④ Looks 87% similar",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "md",
              "spacing": "sm",
              "type": "box",
            },
            "hero": Object {
              "size": "full",
              "type": "image",
              "url": "http://foo/image2.jpeg",
            },
            "type": "bubble",
          },
          Object {
            "body": undefined,
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":\\"image-article-1\\",\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_ARTICLE\\"}",
                    "displayText": "I choose ⑤",
                    "label": "Choose this one",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "⑤ Looks 100% similar",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "md",
              "spacing": "sm",
              "type": "box",
            },
            "hero": Object {
              "size": "full",
              "type": "image",
              "url": "http://foo/image.jpeg",
            },
            "type": "bubble",
          },
          Object {
            "body": undefined,
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":\\"image-article-2\\",\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_ARTICLE\\"}",
                    "displayText": "I choose ⑥",
                    "label": "Choose this one",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "⑥ Looks 87% similar",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "md",
              "spacing": "sm",
              "type": "box",
            },
            "hero": Object {
              "size": "full",
              "type": "image",
              "url": "http://foo/image2.jpeg",
            },
            "type": "bubble",
          },
          Object {
            "body": undefined,
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":\\"image-article-1\\",\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_ARTICLE\\"}",
                    "displayText": "I choose ⑦",
                    "label": "Choose this one",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "⑦ Looks 100% similar",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "md",
              "spacing": "sm",
              "type": "box",
            },
            "hero": Object {
              "size": "full",
              "type": "image",
              "url": "http://foo/image.jpeg",
            },
            "type": "bubble",
          },
          Object {
            "body": undefined,
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":\\"image-article-2\\",\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_ARTICLE\\"}",
                    "displayText": "I choose ⑧",
                    "label": "Choose this one",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "⑧ Looks 87% similar",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "md",
              "spacing": "sm",
              "type": "box",
            },
            "hero": Object {
              "size": "full",
              "type": "image",
              "url": "http://foo/image2.jpeg",
            },
            "type": "bubble",
          },
          Object {
            "body": undefined,
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":\\"image-article-1\\",\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_ARTICLE\\"}",
                    "displayText": "I choose ⑨",
                    "label": "Choose this one",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "⑨ Looks 100% similar",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "md",
              "spacing": "sm",
              "type": "box",
            },
            "hero": Object {
              "size": "full",
              "type": "image",
              "url": "http://foo/image.jpeg",
            },
            "type": "bubble",
          },
        ],
        "type": "carousel",
      },
      "type": "flex",
    },
  ],
}
`;
